File: backend/old/src/services/adapterConfigService.ts
/**
 * QQ适配器配置服务
 * 提供对QQ适配器config.toml文件的读取和写入操作
 */

import { logger } from '../core/logger';
import { pathCacheManager } from '../core/pathCacheManager';
import { 
  AdapterConfig,
  AdapterConfigUpdateData, 
  AdapterConfigFileInfo, 
  AdapterConfigServiceOptions 
} from '../types/adapterConfig';
import { ValidationError, NotFoundError, InternalServerError } from '../middleware/errorHandler';
import { stringifyConfigToCleanToml } from '../utils/tomlHelpers';
import * as TOML from '@iarna/toml';
import * as fs from 'fs';
import * as path from 'path';

export class AdapterConfigService {
  private static readonly CONFIG_FILE_NAME = 'config.toml';
  private static readonly ADAPTER_NAME = 'QQ适配器';
  private static readonly DEFAULT_OPTIONS: AdapterConfigServiceOptions = {
    encoding: 'utf-8',
    backup: true,
    validate: true
  };

  /**
   * 获取配置文件的完整路径
   */
  private static getConfigPath(): string {
    const adapterRoot = pathCacheManager.getAdapterRoot(this.ADAPTER_NAME);
    if (!adapterRoot) {
      throw new NotFoundError('QQ适配器根目录未设置，请先在路径缓存中设置适配器根目录');
    }

    return path.join(adapterRoot, this.CONFIG_FILE_NAME);
  }

  /**
   * 检查配置文件信息
   */
  public static async getConfigFileInfo(): Promise<AdapterConfigFileInfo> {
    try {
      const configPath = this.getConfigPath();
      
      let stats: fs.Stats | null = null;
      let exists = false;
      
      try {
        stats = await fs.promises.stat(configPath);
        exists = true;
      } catch (error) {
        if ((error as NodeJS.ErrnoException).code !== 'ENOENT') {
          throw error;
        }
      }

      const info: AdapterConfigFileInfo = {
        path: configPath,
        exists,
        readable: false,
        writable: false,
        size: 0,
        lastModified: new Date(0)
      };

      if (exists && stats) {
        info.size = stats.size;
        info.lastModified = stats.mtime;

        // 检查读写权限
        try {
          await fs.promises.access(configPath, fs.constants.R_OK);
          info.readable = true;
        } catch {
          info.readable = false;
        }

        try {
          await fs.promises.access(configPath, fs.constants.W_OK);
          info.writable = true;
        } catch {
          info.writable = false;
        }
      }

      return info;
    } catch (error) {
      logger.error('获取QQ适配器配置文件信息失败:', error);
      throw new InternalServerError('获取QQ适配器配置文件信息失败');
    }
  }

  /**
   * 读取QQ适配器配置文件
   */
  public static async getConfig(options: Partial<AdapterConfigServiceOptions> = {}): Promise<AdapterConfig> {
    try {
      const opts = { ...this.DEFAULT_OPTIONS, ...options };
      const configPath = this.getConfigPath();

      // 检查文件是否存在
      const fileInfo = await this.getConfigFileInfo();
      if (!fileInfo.exists) {
        throw new NotFoundError(`QQ适配器配置文件不存在: ${configPath}`);
      }

      if (!fileInfo.readable) {
        throw new InternalServerError(`QQ适配器配置文件不可读: ${configPath}`);
      }

      // 读取文件内容
      let fileContent: string;
      try {
        const content = await fs.promises.readFile(configPath, { encoding: opts.encoding });
        fileContent = typeof content === 'string' ? content : content.toString();
      } catch (error) {
        logger.error('读取QQ适配器配置文件失败:', error);
        throw new InternalServerError(`读取QQ适配器配置文件失败: ${error instanceof Error ? error.message : '未知错误'}`);
      }

      // 解析TOML内容
      let configData: AdapterConfig;
      try {
        configData = TOML.parse(fileContent) as unknown as AdapterConfig;
      } catch (error) {
        logger.error('解析QQ适配器配置文件失败:', error);
        throw new ValidationError(`QQ适配器配置文件格式错误: ${error instanceof Error ? error.message : 'TOML解析失败'}`);
      }

      logger.info('QQ适配器配置读取成功');
      return configData;

    } catch (error) {
      logger.error('获取QQ适配器配置失败:', error);
      throw error instanceof ValidationError || error instanceof NotFoundError || error instanceof InternalServerError
        ? error 
        : new InternalServerError('获取QQ适配器配置失败');
    }
  }

  /**
   * 更新QQ适配器配置文件
   */
  public static async updateConfig(
    updateData: AdapterConfigUpdateData, 
    options: Partial<AdapterConfigServiceOptions> = {}
  ): Promise<void> {
    try {
      const opts = { ...this.DEFAULT_OPTIONS, ...options };
      const configPath = this.getConfigPath();

      if (!updateData || Object.keys(updateData).length === 0) {
        throw new ValidationError('更新数据不能为空');
      }

      // 验证更新数据格式
      if (opts.validate) {
        this.validateConfigData(updateData);
      }

      // 检查文件状态
      const fileInfo = await this.getConfigFileInfo();
      
      let currentConfig: AdapterConfig;
      
      // 如果文件存在，先读取当前配置
      if (fileInfo.exists) {
        if (!fileInfo.readable) {
          throw new InternalServerError(`QQ适配器配置文件不可读: ${configPath}`);
        }
        if (!fileInfo.writable) {
          throw new InternalServerError(`QQ适配器配置文件不可写: ${configPath}`);
        }
        
        currentConfig = await this.getConfig({ validate: false });
      } else {
        // 文件不存在，使用默认配置结构
        throw new NotFoundError(`QQ适配器配置文件不存在: ${configPath}`);
      }

      // 合并配置数据
      const mergedConfig = this.mergeConfig(currentConfig, updateData);

      // 创建备份（如果文件存在且启用备份）
      if (fileInfo.exists && opts.backup) {
        await this.createBackup(configPath);
      }

      // 转换为TOML格式并写入文件
      try {
        const tomlContent = stringifyConfigToCleanToml(mergedConfig, TOML);
        await fs.promises.writeFile(configPath, tomlContent, { encoding: opts.encoding });
      } catch (error) {
        logger.error('写入QQ适配器配置文件失败:', error);
        throw new InternalServerError(`写入QQ适配器配置文件失败: ${error instanceof Error ? error.message : '未知错误'}`);
      }

      logger.info('QQ适配器配置更新成功');

    } catch (error) {
      logger.error('更新QQ适配器配置失败:', error);
      throw error instanceof ValidationError || error instanceof NotFoundError || error instanceof InternalServerError
        ? error 
        : new InternalServerError('更新QQ适配器配置失败');
    }
  }

  /**
   * 验证配置数据格式
   */
  private static validateConfigData(data: AdapterConfigUpdateData): void {
    // 基础类型检查
    if (typeof data !== 'object' || data === null || Array.isArray(data)) {
      throw new ValidationError('配置数据必须是一个对象');
    }

    // 验证特定的配置结构
    if (data.napcat_server) {
      this.validateServerConfig(data.napcat_server, 'napcat_server');
    }

    if (data.maibot_server) {
      this.validateServerConfig(data.maibot_server, 'maibot_server');
    }

    if (data.chat) {
      this.validateChatConfig(data.chat);
    }

    if (data.debug && data.debug.level) {
      const validLevels = ['DEBUG', 'INFO', 'WARNING', 'ERROR', 'CRITICAL'];
      if (!validLevels.includes(data.debug.level)) {
        throw new ValidationError(`无效的日志等级: ${data.debug.level}`);
      }
    }

    // 递归检查嵌套对象，确保数据结构合理
    const validateObject = (obj: any, path = ''): void => {
      for (const [key, value] of Object.entries(obj)) {
        const currentPath = path ? `${path}.${key}` : key;
        
        // 检查键名
        if (typeof key !== 'string' || key.trim() === '') {
          throw new ValidationError(`配置项键名无效: ${currentPath}`);
        }

        // 检查值类型（TOML支持的类型）
        const valueType = typeof value;
        const isValidType = [
          'string', 'number', 'boolean'
        ].includes(valueType) || 
        value === null || 
        Array.isArray(value) || 
        (valueType === 'object' && value !== null);

        if (!isValidType) {
          throw new ValidationError(`配置项类型不支持: ${currentPath} (${valueType})`);
        }

        // 递归验证嵌套对象
        if (valueType === 'object' && value !== null && !Array.isArray(value)) {
          validateObject(value, currentPath);
        }

        // 验证数组元素类型一致性
        if (Array.isArray(value) && value.length > 0) {
          const firstType = typeof value[0];
          const allSameType = value.every(item => typeof item === firstType);
          if (!allSameType) {
            throw new ValidationError(`数组元素类型不一致: ${currentPath}`);
          }
        }
      }
    };

    validateObject(data);
  }

  /**
   * 验证服务器配置
   */
  private static validateServerConfig(serverConfig: any, configName: string): void {
    if (typeof serverConfig !== 'object' || serverConfig === null) {
      throw new ValidationError(`${configName} 必须是一个对象`);
    }

    if (serverConfig.host && typeof serverConfig.host !== 'string') {
      throw new ValidationError(`${configName}.host 必须是字符串`);
    }

    if (serverConfig.port) {
      if (typeof serverConfig.port !== 'number' || serverConfig.port < 1 || serverConfig.port > 65535) {
        throw new ValidationError(`${configName}.port 必须是1-65535之间的数字`);
      }
    }

    if (serverConfig.heartbeat_interval) {
      if (typeof serverConfig.heartbeat_interval !== 'number' || serverConfig.heartbeat_interval < 1) {
        throw new ValidationError(`${configName}.heartbeat_interval 必须是大于0的数字`);
      }
    }
  }

  /**
   * 验证聊天配置
   */
  private static validateChatConfig(chatConfig: any): void {
    if (typeof chatConfig !== 'object' || chatConfig === null) {
      throw new ValidationError('chat 配置必须是一个对象');
    }

    // 验证名单类型
    if (chatConfig.group_list_type && !['whitelist', 'blacklist'].includes(chatConfig.group_list_type)) {
      throw new ValidationError('group_list_type 必须是 whitelist 或 blacklist');
    }

    if (chatConfig.private_list_type && !['whitelist', 'blacklist'].includes(chatConfig.private_list_type)) {
      throw new ValidationError('private_list_type 必须是 whitelist 或 blacklist');
    }

    // 验证列表是否为数字数组
    const validateNumberArray = (arr: any, name: string) => {
      if (!Array.isArray(arr)) {
        throw new ValidationError(`${name} 必须是数组`);
      }
      if (!arr.every(item => typeof item === 'number' && Number.isInteger(item) && item > 0)) {
        throw new ValidationError(`${name} 数组中的所有元素必须是正整数`);
      }
    };

    if (chatConfig.group_list) {
      validateNumberArray(chatConfig.group_list, 'group_list');
    }

    if (chatConfig.private_list) {
      validateNumberArray(chatConfig.private_list, 'private_list');
    }

    if (chatConfig.ban_user_id) {
      validateNumberArray(chatConfig.ban_user_id, 'ban_user_id');
    }

    // 验证布尔值
    if (chatConfig.ban_qq_bot !== undefined && typeof chatConfig.ban_qq_bot !== 'boolean') {
      throw new ValidationError('ban_qq_bot 必须是布尔值');
    }

    if (chatConfig.enable_poke !== undefined && typeof chatConfig.enable_poke !== 'boolean') {
      throw new ValidationError('enable_poke 必须是布尔值');
    }
  }

  /**
   * 合并配置数据（深度合并）
   */
  private static mergeConfig(current: AdapterConfig, update: AdapterConfigUpdateData): AdapterConfig {
    const merged = { ...current };

    for (const [key, value] of Object.entries(update)) {
      if (value !== undefined) {
        if (
          typeof value === 'object' && 
          value !== null && 
          !Array.isArray(value) &&
          typeof merged[key as keyof AdapterConfig] === 'object' && 
          merged[key as keyof AdapterConfig] !== null && 
          !Array.isArray(merged[key as keyof AdapterConfig])
        ) {
          // 递归合并嵌套对象
          (merged as any)[key] = this.mergeConfig((merged as any)[key], value);
        } else {
          // 直接替换值
          (merged as any)[key] = value;
        }
      }
    }

    return merged;
  }

  /**
   * 创建配置文件备份
   */
  private static async createBackup(configPath: string): Promise<void> {
    try {
      const adapterRoot = pathCacheManager.getAdapterRoot(this.ADAPTER_NAME);
      if (!adapterRoot) {
        throw new Error('无法获取QQ适配器根目录路径');
      }

      // 创建备份目录结构：LauncherConfigBak/adapter/qq/
      const backupDir = path.join(path.dirname(adapterRoot), 'config', 'LauncherConfigBak', 'adapter', 'qq');
      await fs.promises.mkdir(backupDir, { recursive: true });

      // 生成带时间戳的备份文件名
      const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
      const configFileName = path.basename(configPath);
      const backupFileName = `${configFileName}.backup.${timestamp}`;
      const backupPath = path.join(backupDir, backupFileName);

      // 复制配置文件到备份目录
      await fs.promises.copyFile(configPath, backupPath);
      logger.info(`QQ适配器配置备份已创建: ${backupPath}`);

    } catch (error) {
      logger.warn('创建QQ适配器配置文件备份失败:', error);
      // 备份失败不应该阻止配置更新，只记录警告
    }
  }
}

File: backend/src/services/adapter_config_service.py
"""
适配器配置服务
Adapter Configuration Service
提供对QQ适配器config.toml文件的读取和写入操作
"""

import os
import shutil
from datetime import datetime
from typing import Optional, Dict, Any

import toml

from core.logger import logger
from core.path_cache_manager import path_cache_manager
from models.config import (
    AdapterConfigData,
    ConfigUpdateData,
    ConfigFileInfo,
    ConfigServiceOptions
)
from utils.toml_helpers import stringify_config_to_clean_toml


class AdapterConfigService:
    """适配器配置服务类"""
    
    CONFIG_FILE_NAME = 'config.toml'
    ADAPTER_NAME = 'QQ适配器'
    
    @classmethod
    def _get_config_path(cls) -> str:
        """获取配置文件的完整路径"""
        adapter_root = path_cache_manager.get_adapter_root(cls.ADAPTER_NAME)
        if not adapter_root:
            raise FileNotFoundError('QQ适配器根目录未设置，请先在路径缓存中设置适配器根目录')
        
        return os.path.join(adapter_root, cls.CONFIG_FILE_NAME)
    
    @classmethod
    async def get_config_file_info(cls) -> ConfigFileInfo:
        """检查配置文件信息"""
        try:
            config_path = cls._get_config_path()
            
            exists = os.path.exists(config_path)
            readable = False
            writable = False
            size = 0
            last_modified = datetime.now().isoformat()
            
            if exists:
                stat_info = os.stat(config_path)
                size = stat_info.st_size
                last_modified = datetime.fromtimestamp(stat_info.st_mtime).isoformat()
                
                # 检查读写权限
                readable = os.access(config_path, os.R_OK)
                writable = os.access(config_path, os.W_OK)
            
            return ConfigFileInfo(
                path=config_path,
                exists=exists,
                readable=readable,
                writable=writable,
                size=size,
                last_modified=last_modified
            )
            
        except Exception as error:
            logger.error(f'获取QQ适配器配置文件信息失败: {error}')
            raise RuntimeError('获取QQ适配器配置文件信息失败')
    
    @classmethod
    async def get_config(cls, options: Optional[ConfigServiceOptions] = None) -> AdapterConfigData:
        """读取QQ适配器配置文件"""
        try:
            opts = options or ConfigServiceOptions()
            config_path = cls._get_config_path()
            
            # 检查文件是否存在
            file_info = await cls.get_config_file_info()
            if not file_info.exists:
                raise FileNotFoundError(f'QQ适配器配置文件不存在: {config_path}')
            
            if not file_info.readable:
                raise PermissionError(f'QQ适配器配置文件不可读: {config_path}')
            
            # 读取文件内容
            try:
                with open(config_path, 'r', encoding=opts.encoding) as f:
                    file_content = f.read()
            except Exception as error:
                logger.error(f'读取QQ适配器配置文件失败: {error}')
                raise RuntimeError(f'读取QQ适配器配置文件失败: {error}')
            
            # 解析TOML内容
            try:
                config_data = toml.loads(file_content)
            except Exception as error:
                logger.error(f'解析QQ适配器配置文件失败: {error}')
                raise ValueError(f'QQ适配器配置文件格式错误: {error}')
            
            logger.info('QQ适配器配置读取成功')
            return AdapterConfigData(**config_data)
            
        except Exception as error:
            logger.error(f'获取QQ适配器配置失败: {error}')
            raise error
    
    @classmethod
    async def update_config(
        cls,
        update_data: ConfigUpdateData,
        options: Optional[ConfigServiceOptions] = None
    ) -> None:
        """更新QQ适配器配置文件"""
        try:
            opts = options or ConfigServiceOptions()
            config_path = cls._get_config_path()
            
            if not update_data:
                raise ValueError('更新数据不能为空')
            
            update_dict = update_data.model_dump()
            if not update_dict:
                raise ValueError('更新数据不能为空')
            
            # 验证更新数据格式
            if opts.validate:
                cls._validate_config_data(update_dict)
            
            # 检查文件状态
            file_info = await cls.get_config_file_info()
            
            if not file_info.exists:
                raise FileNotFoundError(f'QQ适配器配置文件不存在: {config_path}')
            
            if not file_info.readable:
                raise PermissionError(f'QQ适配器配置文件不可读: {config_path}')
            if not file_info.writable:
                raise PermissionError(f'QQ适配器配置文件不可写: {config_path}')
            
            # 读取当前配置
            current_config_data = await cls.get_config(ConfigServiceOptions(validate=False))
            current_config = current_config_data.model_dump()
            
            # 合并配置数据
            merged_config = cls._merge_config(current_config, update_dict)
            
            # 创建备份（如果启用备份）
            if opts.backup:
                await cls._create_backup(config_path)
            
            # 转换为TOML格式并写入文件
            try:
                toml_content = stringify_config_to_clean_toml(merged_config, toml)
                with open(config_path, 'w', encoding=opts.encoding) as f:
                    f.write(toml_content)
            except Exception as error:
                logger.error(f'写入QQ适配器配置文件失败: {error}')
                raise RuntimeError(f'写入QQ适配器配置文件失败: {error}')
            
            logger.info('QQ适配器配置更新成功')
            
        except Exception as error:
            logger.error(f'更新QQ适配器配置失败: {error}')
            raise error
    
    @classmethod
    def _validate_config_data(cls, data: Dict[str, Any]) -> None:
        """验证配置数据格式"""
        if not isinstance(data, dict):
            raise ValueError('配置数据必须是一个对象')
        
        # 验证特定的配置结构
        if 'napcat_server' in data:
            cls._validate_server_config(data['napcat_server'], 'napcat_server')
        
        if 'maibot_server' in data:
            cls._validate_server_config(data['maibot_server'], 'maibot_server')
        
        if 'chat' in data:
            cls._validate_chat_config(data['chat'])
        
        if 'debug' in data and 'level' in data['debug']:
            valid_levels = ['DEBUG', 'INFO', 'WARNING', 'ERROR', 'CRITICAL']
            if data['debug']['level'] not in valid_levels:
                raise ValueError(f'无效的日志等级: {data["debug"]["level"]}')
        
        # 递归检查嵌套对象，确保数据结构合理
        def validate_object(obj: Any, path: str = '') -> None:
            if isinstance(obj, dict):
                for key, value in obj.items():
                    current_path = f'{path}.{key}' if path else key
                    
                    # 检查键名
                    if not isinstance(key, str) or not key.strip():
                        raise ValueError(f'配置项键名无效: {current_path}')
                    
                    # 检查值类型（TOML支持的类型）
                    if not cls._is_valid_toml_type(value):
                        raise ValueError(f'配置项类型不支持: {current_path} ({type(value).__name__})')
                    
                    # 递归验证嵌套对象
                    if isinstance(value, (dict, list)):
                        validate_object(value, current_path)
            
            elif isinstance(obj, list):
                # 验证数组元素类型一致性
                if obj and len(obj) > 0:
                    first_type = type(obj[0])
                    if not all(isinstance(item, first_type) for item in obj):
                        raise ValueError(f'数组元素类型不一致: {path}')
                    
                    # 递归验证数组元素
                    for i, item in enumerate(obj):
                        validate_object(item, f'{path}[{i}]')
        
        validate_object(data)
    
    @classmethod
    def _validate_server_config(cls, server_config: Any, config_name: str) -> None:
        """验证服务器配置"""
        if not isinstance(server_config, dict):
            raise ValueError(f'{config_name} 必须是一个对象')
        
        if 'host' in server_config and not isinstance(server_config['host'], str):
            raise ValueError(f'{config_name}.host 必须是字符串')
        
        if 'port' in server_config:
            if not isinstance(server_config['port'], int) or not (1 <= server_config['port'] <= 65535):
                raise ValueError(f'{config_name}.port 必须是1-65535之间的数字')
        
        if 'heartbeat_interval' in server_config:
            if not isinstance(server_config['heartbeat_interval'], (int, float)) or server_config['heartbeat_interval'] < 1:
                raise ValueError(f'{config_name}.heartbeat_interval 必须是大于0的数字')
    
    @classmethod
    def _validate_chat_config(cls, chat_config: Any) -> None:
        """验证聊天配置"""
        if not isinstance(chat_config, dict):
            raise ValueError('chat 配置必须是一个对象')
        
        # 验证名单类型
        if 'group_list_type' in chat_config and chat_config['group_list_type'] not in ['whitelist', 'blacklist']:
            raise ValueError('group_list_type 必须是 whitelist 或 blacklist')
        
        if 'private_list_type' in chat_config and chat_config['private_list_type'] not in ['whitelist', 'blacklist']:
            raise ValueError('private_list_type 必须是 whitelist 或 blacklist')
        
        # 验证列表是否为数字数组
        def validate_number_array(arr: Any, name: str) -> None:
            if not isinstance(arr, list):
                raise ValueError(f'{name} 必须是数组')
            if not all(isinstance(item, int) and item > 0 for item in arr):
                raise ValueError(f'{name} 数组中的所有元素必须是正整数')
        
        if 'group_list' in chat_config:
            validate_number_array(chat_config['group_list'], 'group_list')
        
        if 'private_list' in chat_config:
            validate_number_array(chat_config['private_list'], 'private_list')
        
        if 'ban_user_id' in chat_config:
            validate_number_array(chat_config['ban_user_id'], 'ban_user_id')
        
        # 验证布尔值
        if 'ban_qq_bot' in chat_config and not isinstance(chat_config['ban_qq_bot'], bool):
            raise ValueError('ban_qq_bot 必须是布尔值')
        
        if 'enable_poke' in chat_config and not isinstance(chat_config['enable_poke'], bool):
            raise ValueError('enable_poke 必须是布尔值')
    
    @classmethod
    def _is_valid_toml_type(cls, value: Any) -> bool:
        """检查值是否是TOML支持的类型"""
        return (
            isinstance(value, (str, int, float, bool)) or
            value is None or
            isinstance(value, list) or
            isinstance(value, dict)
        )
    
    @classmethod
    def _merge_config(cls, current: Dict[str, Any], update: Dict[str, Any]) -> Dict[str, Any]:
        """合并配置数据（深度合并）"""
        merged = current.copy()
        
        for key, value in update.items():
            if value is not None:
                if (
                    isinstance(value, dict) and
                    key in merged and
                    isinstance(merged[key], dict)
                ):
                    # 递归合并嵌套对象
                    merged[key] = cls._merge_config(merged[key], value)
                else:
                    # 直接替换值
                    merged[key] = value
        
        return merged
    
    @classmethod
    async def _create_backup(cls, config_path: str) -> None:
        """创建配置文件备份"""
        try:
            adapter_root = path_cache_manager.get_adapter_root(cls.ADAPTER_NAME)
            if not adapter_root:
                raise RuntimeError('无法获取QQ适配器根目录路径')
            
            # 创建备份目录结构：LauncherConfigBak/adapter/qq/
            backup_dir = os.path.join(os.path.dirname(adapter_root), 'config', 'LauncherConfigBak', 'adapter', 'qq')
            os.makedirs(backup_dir, exist_ok=True)
            
            # 生成带时间戳的备份文件名
            timestamp = datetime.now().isoformat().replace(':', '-').replace('.', '-')
            config_filename = os.path.basename(config_path)
            backup_filename = f'{config_filename}.backup.{timestamp}'
            backup_path = os.path.join(backup_dir, backup_filename)
            
            # 复制配置文件到备份目录
            shutil.copy2(config_path, backup_path)
            logger.info(f'QQ适配器配置备份已创建: {backup_path}')
            
        except Exception as error:
            logger.warning(f'创建QQ适配器配置文件备份失败: {error}')
            # 备份失败不应该阻止配置更新，只记录警告
